triton-cli/src/main.rs

```diff
@@ -0,0 +1,214 @@
+    }
+    for v in claim.input {
+        println!("{v}");
+    }
+    println!("public outputs:");
+    if claim.output.is_empty() {
+        println!("(none)");
+    }
+    for v in claim.output {
+        println!("{v}");
+    }
+    Ok(())
+}
+
+fn parse_inputs(inputs: Option<String>) -> Vec<BFieldElement> {
+    inputs
+        .unwrap_or_default()
+        .split(',')
+        .filter(|v| !v.is_empty())
+        .map(|v| v.parse().unwrap())
+        .collect()
+}
+
+fn prove(
+    asm_filepath: &str,
+    out: &str,
+    public_inputs: Option<String>,
+    private_inputs: Option<String>,
+) -> Result<()> {
+    if std::path::Path::new(out).exists() {
+        anyhow::bail!("output file already exists: {out}");
+    }
+    let asm = fs::read_to_string(asm_filepath)
+        .with_context(|| "Failed to read Triton assembly from file")?;
+    let program = triton_program!({ asm });
+
+    let public_input = PublicInput::from(parse_inputs(public_inputs));
+    let non_determinism = NonDeterminism::from(parse_inputs(private_inputs));
+    println!("proving...");
+    let data = triton_vm::prove_program(&program, public_input, non_determinism)
+        .with_context(|| "Triton VM errored during program execution")?;
+    println!("success!");
+
+    write_proof(data, out).with_context(|| "Failed to write proof to file")?;
+    println!("proof written to: {out}");
+    Ok(())
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct SerializedProof {
+    version: u8,
+    security_level: u64,
+    fri_expansion_factor: u64,
+    num_trace_randomizers: u64,
+    num_colinearity_checks: u64,
+    num_combination_codeword_checks: u64,
+    program_digest: Vec<u64>,
+    public_inputs: Vec<u64>,
+    public_outputs: Vec<u64>,
+    proof: Vec<u64>,
+}
+
+fn read_proof(proof_path: &str) -> Result<(Stark, Claim, Proof)> {
+    let proof_bytes = fs::read(proof_path)?;
+    let serialized_proof: SerializedProof = bincode::deserialize(&proof_bytes)?;
+    if serialized_proof.version != 1 {
+        anyhow::bail!("wrong proof file version!");
+    }
+    let stark = Stark {
+        security_level: usize::try_from(serialized_proof.security_level)?,
+        fri_expansion_factor: usize::try_from(serialized_proof.fri_expansion_factor)?,
+        num_trace_randomizers: usize::try_from(serialized_proof.num_trace_randomizers)?,
+        num_collinearity_checks: usize::try_from(serialized_proof.num_colinearity_checks)?,
+        num_combination_codeword_checks: usize::try_from(
+            serialized_proof.num_combination_codeword_checks,
+        )?,
+    };
+    assert_eq!(
+        DIGEST_LENGTH,
+        serialized_proof.program_digest.len(),
+        "digest length mismatch!"
+    );
+    let digest_elements = serialized_proof
+        .program_digest
+        .iter()
+        .map(|v| BFieldElement::from(*v))
+        .collect::<Vec<BFieldElement>>();
+    let digest = Digest::try_from(&digest_elements[0..DIGEST_LENGTH])?;
```

Comment:
You don't have to do the length check yourself if you pass the entire `digest_elements` into the `try_from`:

```suggestion
    let digest_elements = serialized_proof
        .program_digest
        .iter()
        .map(|v| BFieldElement::from(*v))
        .collect::<Vec<_>>();
    let digest = Digest::try_from(digest_elements)?;
```

Reply:
In fact, you can remove a lot of `.iter().map(…)`ing by changing the types used in the `SerializedProof`:

```rust
#[derive(Serialize, Deserialize, Debug, Clone)]
struct SerializedProof {
	..
    program_digest: Digest,
    public_inputs: Vec<BFieldElement>,
    public_outputs: Vec<BFieldElement>,
    proof: Vec<BFieldElement>,
}
```

All of these types (de)serialize just fine. :slightly_smiling_face:

Reply:
I applied the first change passing the full vector to `try_from`. I thought about changing the types in `SerializedProof`, but this makes the file structure vulnerable to change if `Digest` or `BFieldElement` changes. That's why i'm unwrapping/re-wrapping into primitive types.